In [55]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [56]:
columns = ['Age','Job_type','Final_wt','Education_level','Education_Yr',
           'Marital_status','Job_role','Family_role','Race',
           'Gender','Capital_gain','Capital_loss','Weekly_hrs',
           'Country','Income']

In [57]:
import os
df = pd.read_csv('data/adult.data', names=columns, na_values='?', skipinitialspace=True)
print(df.shape)
df.head()

(32561, 15)


,Age,Job_type,Final_wt,Education_level,Education_Yr,Marital_status,Job_role,Family_role,Race,Gender,Capital_gain,Capital_loss,Weekly_hrs,Country,Income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


In [58]:
# checking missing values 
missing_values = df.isna().sum()
print(missing_values)

Age                   0
Job_type           1836
Final_wt              0
Education_level       0
Education_Yr          0
Marital_status        0
Job_role           1843
Family_role           0
Race                  0
Gender                0
Capital_gain          0
Capital_loss          0
Weekly_hrs            0
Country             583
Income                0
dtype: int64


In [59]:
print("Shape before : ", df.shape)

Shape before :  (32561, 15)


In [60]:
df_clean = df.dropna()

In [61]:
print("Shape before cleaning : ",df.shape)
print("Shape after cleaning :", df_clean.shape )

Shape before cleaning :  (32561, 15)
Shape after cleaning : (30162, 15)


In [62]:
print("Rows remove : ",df.shape[0] -  df_clean.shape[0])

Rows remove :  2399


In [63]:
print("No. of rows and col in clean dataset : ",df_clean.shape)

No. of rows and col in clean dataset :  (30162, 15)


In [64]:
df= df_clean

In [65]:
df.head()

,Age,Job_type,Final_wt,Education_level,Education_Yr,Marital_status,Job_role,Family_role,Race,Gender,Capital_gain,Capital_loss,Weekly_hrs,Country,Income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


In [67]:
df["Income"].value_counts()

Income
<=50K    22654
>50K      7508
Name: count, dtype: int64

In [68]:
df["Income"].value_counts(normalize=True)*100

Income
<=50K    75.107751
>50K     24.892249
Name: proportion, dtype: float64

We can say data is imbalanced since nearly 75% are earning <=50k which can lead to biased model predictions toward the majority class.

In [69]:
df["Gender"].value_counts()

Gender
Male      20380
Female     9782
Name: count, dtype: int64

In [73]:
pd.crosstab(df["Gender"], df["Income"],normalize="index")*100

Income,<=50K,>50K
Gender,,
Female,88.632182,11.367818
Male,68.616290,31.383710


The dataset exhibits strong gender bias, where only ~11% of females earn >50K compared to ~31% of males, indicating significant disparity in outcomes.

In [74]:
pd.crosstab(df["Race"], df["Income"], normalize="index")*100

Income,<=50K,>50K
Race,,
Amer-Indian-Eskimo,88.111888,11.888112
Asian-Pac-Islander,72.290503,27.709497
Black,87.007455,12.992545
Other,90.909091,9.090909
White,73.628196,26.371804


The dataset shows significant racial bias, where Asian-Pac-Islander (~27.7%) and White (~26.3%) groups have a much higher proportion of individuals earning >50K compared to groups like Black (~13%), Amer-Indian-Eskimo (~11.8%), and Other (~9%).

In [76]:
df.to_csv("data/clean_data.csv", index=False)